Create new v2_to_v2 integration test

benalleng opened the pull request:
Creates new v2_to_v2 integration test

benalleng committed:
Introduce bitcoin_ffi Psbt struct and test_utils runtime

We need more control over psbts in the integration tests so
I am using my branch of bitcoin_ffi until we have that PR merged
into main.

arminsabouri committed:
Expose uniffi persistence traits as foreign

Callback methods meant to be overridden by the foreign languange should
be exposed with
[`with_foreign`](https://mozilla.github.io/uniffi-rs/0.27/foreign_traits.html#1-define-a-rust-trait)

benalleng commented:
One thing I think I could improve is the ffi `call` method to allow it to accept something besides json so There isn't all the `json.dumps` => `json.loads` clutter. but other than that I am happy with where this is at now!

benalleng committed:
Add the python integration test to workflow

Now that the test is working we should add it to the workflow. I have
left this as a seperate commit in case it adds too much time to the
workflows.

By adding ubuntu latest for linux and a special workflow command for
macos we are able to enable redis to create a compatible workflow for
running the new integration tests.

0xBEEFCAF3 commented on python/test/payjoin_integration_test.py:
```diff
@@ -25,212 +29,237 @@
-            )["psbt"]
+            receiver_address = bitcoinffi.Address(json.loads(self.receiver.call("getnewaddress", [None, None])), bitcoinffi.Network.REGTEST)
+            init_tracing()
+            self.receiver.call("generatetoaddress", [json.dumps(101), json.dumps("mz3QWUnL94q2RdRxFtPq747SngoBu5uQMi")])
```
Can you remind me whats going on here? Do we want to be generting to the receiver address?

0xBEEFCAF3 commented on python/test/payjoin_integration_test.py:
```diff
@@ -25,212 +29,237 @@
+
+            # **********************
+            # Inside the Receiver:
+            expiry: Optional[int] = None
+            new_receiver = NewReceiver(receiver_address, directory.as_string(), ohttp_keys, expiry)
```
Uneccecary expiry var. We can just use None.

0xBEEFCAF3 commented on python/test/payjoin_integration_test.py:
```diff
@@ -25,212 +29,237 @@
+                headers={"Content-Type": request.request.content_type},
+                content=request.request.body
+            )
+            # print(f"Response: {response}")
```
nit: rm

0xBEEFCAF3 commented on python/test/payjoin_integration_test.py:
```diff
@@ -25,212 +29,237 @@
+            # POST payjoin
+            proposal = session.process_res(response.content, request.client_response)
+            payjoin_proposal = handle_directory_payjoin_proposal(self.receiver, proposal)
+            # print(f"Proposal psbt: {payjoin_proposal.psbt()}")
```
nit: rm

0xBEEFCAF3 commented on python/test/payjoin_integration_test.py:
```diff
@@ -25,212 +29,237 @@
+            final_psbt = json.loads(self.sender.call("finalizepsbt", [payjoin_psbt, json.dumps(False)]))["psbt"]
+            payjoin_tx = bitcoinffi.Psbt.deserialize_base64(final_psbt).extract_tx()
+            self.sender.call("sendrawtransaction", [json.dumps(payjoin_tx.serialize().hex())])
+            # print(f"Tx sent: {payjoin_tx.compute_txid()}")
```
nit: rm

0xBEEFCAF3 commented on python/test/payjoin_integration_test.py:
```diff
@@ -25,212 +29,237 @@
+            sequence=0,
+            witness=[]
+        )
+        raw_tx_hex = json.loads(rpc_connection.call("gettransaction", [json.dumps(utxo["txid"]), json.dumps(True)]))["hex"]
```
Nit: I think if you pass True for the verbose flag you should have the decoded tx. And then you don't need to call decoderawtransaction

benalleng replied:
Nice catch, its annoying that `getrawtransaction` has the verbose flag as the second arg and `gettransaction` has it as the 3rd

0xBEEFCAF3 commented on python/test/payjoin_integration_test.py:
```diff
@@ -25,212 +29,237 @@
+                headers={"Content-Type": request.request.content_type},
+                content=request.request.body
+            )
+            # print(f"Response: {response}")
```
nit: rm

0xBEEFCAF3 commented on python/test/payjoin_integration_test.py:
```diff
@@ -25,212 +29,237 @@
+def handle_directory_payjoin_proposal(receiver: Proxy, proposal: UncheckedProposal) -> PayjoinProposal:
+    maybe_inputs_owned = proposal.check_broadcast_suitability(None, MempoolAcceptanceCallback(receiver))
+    maybe_inputs_seen = maybe_inputs_owned.check_inputs_not_owned(IsScriptOwnedCallback(receiver))
+    outputs_unknown = maybe_inputs_seen.check_no_inputs_seen_before(IdentifyReceiverOutputsCallback(receiver))
```
Shouldn't this class `IdentifyReceiverOutputsCallback`  be called `CheckInputsNotSeenCallback` ?

arminsabouri reviewed (commented):
Thanks for pushing this through the finish line! My comments are mostly Nit. The biggest comment is around wether we wait for your upstream PR to merge before merging this PR.

Tested as of 99726cc3e8951ad938e60ba3afbb8b53cba5872a :heavy_check_mark:

benalleng commented on Cargo.toml:
```diff
@@ -23,21 +23,22 @@ uniffi = { version = "0.29.1", features = ["build"] }
 [dependencies]
 base64 = "0.22.1"
-bitcoin-ffi = { git = "https://github.com/bitcoindevkit/bitcoin-ffi.git", rev = "6b1d131" }
+bitcoin-ffi = { git = "https://github.com/benalleng/bitcoin-ffi.git", rev = "8e3a23b" }
```
I think that's a @DanGould question, based on the state of review in that PR it seems likely that the only thing that will change is how errors are built and presented.

DanGould replied:
some value in that tracking the rest of the ecosystem lets us merge projects when we get there, but I think we're a long way off and we won't stray too far. 

I don't think bdk-ffi will depend on payjoin-ffi as an option any time soon, which would be the reason to stay on the same version of bitcoin-ffi afaict

benalleng commented on Cargo.toml:
```diff
@@ -23,21 +23,22 @@ uniffi = { version = "0.29.1", features = ["build"] }
 serde_json = "1.0.128"
 thiserror = "1.0.58"
 tokio = { version = "1.38.0", features = ["full"], optional = true }
+bitcoind = { version = "0.36.0", features = ["0_21_2"], optional = true }
```
Since we build `_test-utils` with the `init_bitcoind_sender_receiver` we need access to this under that flag

benalleng committed:
Introduce python integration tests v2_to_v2

This adds updated integration tests to work with 0.23 and beyond as it
now utilizes the updated payjoin_test_utils.

benalleng commented:
Updated, rebased, and reorderd to d26d647 so that each commit builds and tests logically.

0xBEEFCAF3 commented:
Regarding the failing CI
>Caught: Failed to execute docker command: Os { code: 2, kind: NotFound, message: "No such file or directory" }

is this b/c we can't find the RPC cookie in the right location?

benalleng commented on Cargo.toml:
```diff
@@ -6,7 +6,7 @@ license = "MIT OR Apache-2.0"
 
 [features]
-_test-utils = ["payjoin-test-utils", "tokio"]
+_test-utils = ["payjoin-test-utils", "tokio", "bitcoind"]
```
Wouldn't I still need a reliable way to access and pass the correct types across the ffi layer to bitcoind?

spacebear21 replied:
Good point. "payjoin-test-utils re-exporting bitcoind" sounds like a good idea to me, especially since that specific dependency [may change in the future](https://github.com/payjoin/rust-payjoin/issues/369).

benalleng committed:
rename test files

Using filenames that start with test will allow us to run a more
seamless workflow.

benalleng commented:
Ok, I have drilled down to the issue, If I understand correctly the `many_linux` image is the root cause as it does not have access to the same environments by default as `ubunut-latest` I was able to get past the docker not found by adding a docker cli step but It is still failing with some redis issues. 
```
2025-05-09T16:54:45.016838Z ERROR payjoin_directory: Redis error: Connection refused (os error 111)
2025-05-09T16:54:45.016858Z ERROR payjoin_directory: Internal server error: Internal server error
...
Caught: SessionError()
```
I am stepping away for the weekend but I am pretty confident this can be fixed with a more complete workflow environment.

DanGould commented:
>  it does not have access to the same environments by default as `ubunut-latest`

In case this build fails (Success pls???)

Is there a good reason not to use `ubuntu-latest`?

benalleng commented:
> > it does not have access to the same environments by default as `ubunut-latest`
> 
> In case this build fails (Success pls???)
> 
> Is there a good reason not to use `ubuntu-latest`?

Not really, I was going to try that soon if this current attempt doesn't work, I was just hoping to keep the minimal manylinux image if possible, it might be but perhaps its not worth it

DanGould commented on src/test_utils.rs:
```diff
@@ -43,56 +134,82 @@ impl From<TestServices> for payjoin_test_utils::TestServices {
 
-    pub async fn fetch_ohttp_keys(&self) -> Result<crate::OhttpKeys, crate::io::IoError> {
-        self.0.lock().await.fetch_ohttp_keys().await.map_err(Into::into).map(Into::into)
+    pub fn fetch_ohttp_keys(&self) -> Result<crate::OhttpKeys, crate::io::IoError> {
+        let runtime = RUNTIME.lock().expect("Lock should not be poisoned");
+        runtime.block_on(async {
+            self.0.lock().await.fetch_ohttp_keys().await.map_err(Into::into).map(Into::into)
+        })
```
The repeated let runtime =, runtime.block_on makes me think this could be dried up by a delarative macro. But there's no rush on this, it can get in after this is merged, just a prettier way to write.

DanGould commented on python/requirements-dev.txt:
```diff
@@ -1,4 +1,4 @@
 toml==0.10.2
 yapf==0.43.0
-
+httpx==0.28.1
```
very pleased we're able to use httpx here instead of exposing rust's reqwest. I'm curious, though not particularly eager, to see if the native http clients for some foreign language can replace fetch_ohttp_keys so we don't have to ship a whole rust TLS stack with our binaries, too.

DanGould reviewed (commented):
cACK 1364c24a3b159cf160600b30d03c2ffc086bffa3

Does CI work on commits before the one I cACKed? Or would there be errors if not for the CI changes in the final commit?  That's my only real concern assuming integration testing is operating as I imagine it is. The other comments are just notes for the future.

DanGould commented:
> [021dab3](https://github.com/LtbLightning/payjoin-ffi/pull/85/commits/021dab377e5598a31ddb30f9c3d03441ad676f3f) causes a CI falure as the integration test has been added to the workflow there but the followup changes to add support redis are not in until [1364c24](https://github.com/LtbLightning/payjoin-ffi/pull/85/commits/1364c24a3b159cf160600b30d03c2ffc086bffa3)

shall those commits then be squashed or tests ignored so that each commit in turn passes CI and git bisect works?

benalleng commented:
I think we should just squahs them such that the integration test doesn't get added to the Ci until the redis compatibility is also included as 021dab3 really doesn't do anything other than cause a CI failure with the new tests added

DanGould commented:
How come the mac version is unable to use redis provided by payjoin-test-utils?

DanGould reviewed (approved):
tACK 83359b5bed60407a10c344d2e978a8de1eb49819

I ran the python test on my local machine.

I'm still confused why redis setup is re-scripted in the CI `yml` though. Why not use the payjoin-test-utils docker version so that it's the same as rust? Perhaps this needs to be done in a follow up.

DanGould commented:
Hmmm but it also seems like there might be a silent test failure in the python tests. Look at the 'Run tests' section

```console
 OHTTP relay binding to port [::]:33681
2025-05-13T17:02:36.074649Z ERROR ohttp_relay::bootstrap::connect: server io error: Transport endpoint is not connected (os error 107)
test_integration_v2_to_v2 (test.test_payjoin_integration_test.TestPayjoin.test_integration_v2_to_v2) ... Executing <Task pending name='Task-2' coro=<TestPayjoin.test_integration_v2_to_v2() running at /home/runner/work/payjoin-ffi/payjoin-ffi/python/test/test_payjoin_integration_test.py:92> cb=[_run_until_complete_cb() at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/base_events.py:181] created at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/runners.py:100> took 3.964 seconds
2025-05-13T17:02:38.449256Z ERROR payjoin_directory: Error serving connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Executing <Task finished name='Task-2' coro=<TestPayjoin.test_integration_v2_to_v2() done, defined at /home/runner/work/payjoin-ffi/payjoin-ffi/python/test/test_payjoin_integration_test.py:71> result=None created at /opt/hostedtoolcache/Python/3.13.3/x64/lib/python3.13/asyncio/runners.py:100> took 0.127 seconds
ok
```

benalleng commented:
> some loop that's waiting and errors the first go around?

I'm pretty sure its the latter as the test still resolves to the end to an ok response, and at least one of these are the same errors on the rust-payjoin.
```
Database running on 127.0.0.1:32865
Directory server binding to port [::]:45329
OHTTP relay binding to port [::]:37679
session: Receiver {
    context: SessionContext {
        address: BCRT1QGZLFDFL2XG56AD00PYEAEE4H9LPWRZ69EQ5WMV,
        directory: Url {
            scheme: "https",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: Some(
                Domain(
                    "localhost",
                ),
            ),
            port: Some(
                45329,
            ),
            path: "/",
            query: None,
            fragment: None,
        },
        subdirectory: None,
        ohttp_keys: OhttpKeys(
            KeyConfig {
                key_id: 1,
                kem: K256Sha256,
                symmetric: [
                    SymmetricSuite {
                        kdf: HkdfSha256,
                        aead: ChaCha20Poly1305,
                    },
                ],
                sk: None,
                pk: PublicKey 04e0f5474a6f2ba35c23862560c7b780de8657a1014ba236e9493cd6aded3df293ee8a2d5ef3899
dd1e6f495c8226c0e1ebadbafcb07c5d49d6f5e5960caa8870f,
            },
        ),
        expiry: SystemTime {
            tv_sec: 1747244986,
            tv_nsec: 249582664,
        },
        s: HpkeKeyPair(
            SecpHpkeSecretKey([REDACTED]),
            SecpHpkePublicKey(PublicKey(PublicKey(e7f5e9d2bf852fdc236f21c2feca061447a6625b754af2b3de0efd82f4e
be4069a8272910cf9705f6259feb35240e378ed16864b879e021d92a905ba48649681))),
        ),
        e: None,
    },
}
2025-05-13T17:49:46.254400Z ERROR ohttp_relay::bootstrap::connect: server io error: Transport endpoint is not
 connected (os error 107)
test integration::v2::v2_to_v2 ... ok
```
I agree though that it would be nice to not get these errors though
